0011-rendezvous-hashing.md

```diff
@@ -0,0 +1,311 @@
+      hashInitialization: 11946695773637837490,
+      shards: [
+        {
+          backend: {
+            grpc: {
+              address: 'storage-0:8980',
+            },
+          },
+          weight: 1,
+        },
+        {
+          backend: {
+            grpc: {
+              address: 'storage-1:8980',
+            },
+          },
+          weight: 1,
+        },
+      ],
+    },
+  },
+  ...
+}
+```
+
+We first want to non-destructively represent the exact same topology with the
+new configuration.
+
+```jsonnet
+{
+  ...
+  contentAddressableStorage: {
+    sharding: {
+      shards: {
+        "0": {
+          backend: { grpc: { address: 'storage-0:8980' } },
+          weight: 1,
+        },
+        "1": {
+          backend: { grpc: { address: 'storage-1:8980' } },
+          weight: 1,
+        },
+      },
+      legacy: {
+        shardOrder: ["0", "1"],
+        hashInitialization: 11946695773637837490,
+      },
+    },
+  },
+  ...
+}
+```
+
+The legacy parameter will make buildbarn internally use the old sharding
+algorithm, the key can be any arbitrary string value. We can then migrate by
+deploying a configuration that uses the new sharding algorithm as a primary but
+falls back to the old sharding algorithm.
+
+```jsonnet
+{
+  ...
+  contentAddressableStorage: {
+    readFallback: {
+      primary: {
+        sharding: {
+          shards: {
+            "0": {
+              backend: { grpc: { address: 'storage-0:8980' } },
+              weight: 1,
+            },
+            "1": {
+              backend: { grpc: { address: 'storage-1:8980' } },
+              weight: 1,
+            },
+          },
+        },
+      },
+      secondary: {
+        sharding: {
+          shards: {
+            "0": {
+              backend: { grpc: { address: 'storage-0:8980' } },
+              weight: 1,
+            },
+            "1": {
+              backend: { grpc: { address: 'storage-1:8980' } },
+              weight: 1,
+            },
+          },
+          legacy: {
+            shardOrder: ["0", "1"],
+            hashInitialization: 11946695773637837490,
+          },
+        },
+      },
+      replicator: { ... }
+    },
+  },
+  ...
+}
+```
+
+When the cluster has used the read fallback configuration for an acceptable
+amount of time you can drop the fallback and have only the new configuration.
+
+```jsonnet
+{
+  ...
+  contentAddressableStorage: {
+    sharding: {
+      shards: {
+        "0": {
+          backend: { grpc: { address: 'storage-0:8980' } },
+          weight: 1,
+        },
+        "1": {
+          backend: { grpc: { address: 'storage-1:8980' } },
+          weight: 1,
+        },
+      },
+    },
+  },
+  ...
+}
+```
+
+## Implementation
+
+Rendezvous hashing was chosen for it's simpleness of implementation,
+mathematical elegance and not needing to be parameterized in configuration.
+
+> Of all shards pick the one with the largest `-weight / log(h)` where `h` is a hash
```

Comment:
Can you use "$" (math mode) for expressions like these?

$\frac{-weight}{\log(h)}$